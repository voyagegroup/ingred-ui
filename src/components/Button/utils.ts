import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
