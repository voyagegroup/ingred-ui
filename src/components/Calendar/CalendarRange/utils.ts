import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
