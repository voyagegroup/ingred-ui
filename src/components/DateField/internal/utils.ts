export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
