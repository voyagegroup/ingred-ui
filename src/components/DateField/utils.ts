import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
