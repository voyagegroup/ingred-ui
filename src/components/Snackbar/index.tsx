export { default } from "./Snackbar";
export type { SnackbarProps } from "./Snackbar";
