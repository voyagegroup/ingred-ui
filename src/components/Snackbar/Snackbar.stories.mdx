import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import Snackbar from "./Snackbar";
import Button from "../Button";

<Meta
  title="Components/Feedback/Snackbar"
  component={Snackbar}
  parameters={{ docs: { source: { type: "code" } } }}
/>

# Snackbar

It express `position: fixed;` panel.

<ArgsTable of={Snackbar} />

## Samples

<Canvas>
  <Story
    name="Example"
    args={{
      isOpen: false,
    }}
  >
    {(args) => {
      const [isOpen, setIsOpen] = React.useState(args.isOpen);
      return (
        <>
          <Button inline={true} onClick={() => setIsOpen(true)}>
            OPEN SNACKBAR
          </Button>
          <Snackbar
            {...args}
            isOpen={isOpen}
            color="default"
            anchorOrigin={{
              vertical: "top",
              horizontal: "left",
            }}
            onClose={() => setIsOpen(false)}
          >
            This is Snackbar(default color).
          </Snackbar>
          <Snackbar
            {...args}
            isOpen={isOpen}
            color="dark"
            anchorOrigin={{
              vertical: "top",
              horizontal: "right",
            }}
            onClose={() => setIsOpen(false)}
          >
            This is Snackbar(dark color).
          </Snackbar>
          <Snackbar
            {...args}
            isOpen={isOpen}
            color="warning"
            anchorOrigin={{
              vertical: "bottom",
              horizontal: "right",
            }}
            onClose={() => setIsOpen(false)}
          >
            This is Snackbar(warning color).
          </Snackbar>
        </>
      );
    }}
  </Story>
</Canvas>
