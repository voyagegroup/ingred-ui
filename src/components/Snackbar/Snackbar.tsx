import * as React from "react";
import * as Styled from "./styled";
import { SnackbarContent } from "./internal/SnackbarContent";

export type SnackbarColor = "default" | "dark" | "warning";

export type SnackbarProps = {
  color?: SnackbarColor;
  /**
   * Define position of this component.
   */
  anchorOrigin?: {
    horizontal: "left" | "right";
    vertical: "top" | "bottom";
  };
  /**
   * If `false`, Snackbar doesn't rendered.
   */
  isOpen?: boolean;
  /**
   * Callback function that triggered when clicked ✖︎ Icon.
   */
  onClose?: () => void;
  children?: React.ReactNode;
};

const Snackbar = React.forwardRef<HTMLDivElement, SnackbarProps>(
  (
    {
      color = "default",
      anchorOrigin: { vertical, horizontal } = {
        vertical: "bottom",
        horizontal: "right",
      },
      isOpen,
      onClose,
      children,
    },
    ref,
  ) => {
    if (!isOpen) {
      return null;
    }
    return (
      <Styled.Container ref={ref} vertical={vertical} horizontal={horizontal}>
        <SnackbarContent color={color} onClose={onClose}>
          {children}
        </SnackbarContent>
      </Styled.Container>
    );
  },
);

export default Snackbar;
