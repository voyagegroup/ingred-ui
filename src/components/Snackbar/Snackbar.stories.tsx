import React from "react";
import { StoryObj } from "@storybook/react";
import { Title, Description, ArgsTable, Stories } from "@storybook/addon-docs";
import Snackbar, { SnackbarProps } from "./Snackbar";
import Button from "../Button";

export default {
  title: "Components/Feedback/Snackbar",
  component: Snackbar,
  args: {
    isOpen: false,
    anchorOrigin: {
      vertical: "top",
      horizontal: "left",
    },
  },
  parameters: {
    docs: {
      source: { language: "tsx" },
      page: () => (
        <>
          <Title />
          <Description markdown={"It express `position: fixed;` panel."} />
          <ArgsTable of={Snackbar} />
          <Stories includePrimary title="Stories" />
        </>
      ),
    },
  },
};

const Template: StoryObj<SnackbarProps> = {
  render: (args) => {
    const [isOpen, setIsOpen] = React.useState(args.isOpen);

    return (
      <>
        <Button inline={true} onClick={() => setIsOpen(true)}>
          OPEN SNACKBAR
        </Button>
        <Snackbar {...args} isOpen={isOpen} onClose={() => setIsOpen(false)}>
          This is Snackbar({args.color} color).
        </Snackbar>
      </>
    );
  },
};

export const Default = {
  ...Template,
  args: {
    color: "default",
  },
};

export const Dark = {
  ...Template,
  args: {
    color: "dark",
  },
};

export const Warning = {
  ...Template,
  args: {
    color: "warning",
  },
};
