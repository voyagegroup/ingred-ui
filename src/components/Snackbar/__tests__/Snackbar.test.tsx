import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import Snackbar from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Snackbar component testing", () => {
  afterEach(cleanup);

  test("Snackbar", () => {
    const { asFragment } = renderWithThemeProvider(<Snackbar isOpen={true} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
