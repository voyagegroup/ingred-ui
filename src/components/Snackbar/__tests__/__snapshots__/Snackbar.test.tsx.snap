// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar component testing Snackbar 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL Gwvzb"
  >
    <div
      class="sc-gsDKAQ jzevRs"
    >
      <div
        class="sc-dkPtRN cDMOpe"
      >
        <div
          class="sc-hKwDye cPkzPO"
          size="18"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#041C33"
              height="23.141"
              rx="2.571"
              transform="translate(18.364 2) rotate(45)"
              width="5.142"
            />
            <rect
              fill="#041C33"
              height="23.141"
              rx="2.571"
              transform="translate(2 5.636) rotate(-45)"
              width="5.143"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
